text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,657 @@
+
+A "Deliver" span covers the call of such a callback or handler and should link
+to the creation context of all messages that are forwarded via the respective
+call.
+
+#### Instrumenting pull-based scenarios
+
+In pull-based consumer scenarios, the delivery of messages is requested by the
+application code. This usually involves a blocking call, which returns zero or
+more messages on completion.
+
+A "Receive" span covers such calls and should link to the creation context of
+all messages that are forwarded via the respective call.
+
+#### General considerations for both push-based and pull-based scenarios
+
+The operations modelled by "Deliver" or "Receive" spans do not strictly refer
+to receiving the message from intermediaries, but instead refer to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
+might pre-fetch messages from intermediaries and cache those messages, and only
+forward messages to the application at a later time. In this case, the
+operation of pre-fetching and caching should not be covered by the "Deliver" or
+"Receive" spans.
+
+Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
+to notify the application that no message is available for processing), one
+message, or multiple messages (a batch of messages). "Deliver" and "Receive"
+spans should link to the creation context of the messages forwarded, thus those
+spans can link to zero, one, or multiple producer spans.
+
+For single-message scenarios, and if the "Deliver" or "Receive" spans would be
+root spans of a new trace, the creation context may also be used as a parent on
+those operations in addition to being added as a link. Keeping single-messages
+operations in the same trace can greatly improve the user experience.
+
+#### Settlement
+
+Messages can be settled in a variety of different ways:
+
+* The intermediary settles the messages as it is sent to the consumer. No
+  settlement operations happen on the consumer.
+* The consumer settles a message without awaiting an acknowledgment from the
+  intermediary.
+* The consumer settles a message and awaits an acknowledgment from the
+  intermediary. This involves a round-trip exchange between the consumer and
+  the intermediary.
+
+Settlement operations on the consumer can either be triggered manually by the
+user, or can be triggered automatically by messaging SDKs based on return
+values of callbacks.
+
+A "Settle" span should be created for every settlement operation that happens
+on the consumer (at-least-once and exactly-once). SDKs will, in some cases,
+auto-settle messages in push-scenarios when messages are delivered via
+callbacks. In cases where it is possible, it is recommended to create the
+"Settle" span as a child of the "Deliver" span.
```

Comment:
I know this is a tired topic in WG discussions, and I think it's much more well-defined now. However, I just want to make sure my last remaining concern here is understood, and that we collectively feel we're making the right trade-off.

My thoughts above are with the following scenario in mind:
* Single message receive
* Only consumer (no intermediary) involved in settlement.
* Auto-ack or client ack from within delivery callback.

I believe this is a very common scenario. It represents one of the simplest messaging scenarios possible (where clients are involved in settlement at all), and why I want to draw attention to it.

The cases where it is possible to create "Settle" as a child of "Deliver" would be in the described use case above.

When the message is settled without waiting for a response from the intermediary (what the AMQP 1.0 specification describes as at least once delivery), I don't think the "end timestamp" provides useful information. The most useful information is when the application performed the settlement. Furthermore, for single message delivery, there's no need to correlate the "settle" with the message creation context since the "Deliver" already does that. IMO this is a very common messaging pattern and is worth considering. There's a trade-off to be made here:
* Create an additional span for the settle operation so that we are consistent and all messages that involve the consumer in settlement always have a settle span. Bear in mind this also means creating a link to the creation context, which is yet more overhead; vs.
* Just create an event in the "Deliver" span with the timestamp the application settled the message. The span already has a link to the creation context.

So ultimately what I'm wondering is whether the value of "every consumer-initiated settlement results in a settle span" worth the cost of providing little other useful information in what I'd consider one of the simplest messaging scenarios where consumers are involved in settlement.

I suppose the normative language would technically allow to place information related to settlement into a delivery span event -- the spec only says that:
> "Settle" spans SHOULD be created for every manually or automatically triggered settlement operation. 

Since SHOULD is used rather than MUST, it does leave it open to omitting the send span and instead including relevant settlement data in a "Deliver" span. But I think an implementation violating a "SHOULD" recommendation would generally be frowned upon, yet I think there's a good justification for omitting the send span in the case I describe above.

Reply:
Let me turn this around a bit and add a third option that would probably work for both points:

- if there is a remote call to settle, it must be a span (to have duration, status, provide scope for transport spans underneath). It comes with a MUST/SHOULD
- if there is no remote call to settle (in memory settlement?), instrumentation (or system-specific messaging convention extension) MAY create a span or MAY instead populate attributes on the deliver span (using certain not-yet-defined attributes to express settlement kind - ack/nack/whatnot)

Reply:
This was discussed in today's workgroup meeting.

We decided to leave the wording as it is, and have a SHOULD requirement for a settlement span for each settlement operation.

If this doesn't work well for individual cases, they can still omit the settlement span (as it's just a SHOULD requirement).

If this turns out to be a general problem and lots of users are unhappy with it, we can re-discuss and maybe change from SHOULD to MAY. We also want to revisit this topic in the context of settlement attribute discussions.